Compiler error with struct references: "member reference type is a pointer; did you mean to use '->'

### Describe the bug

Compiling the file below results in three errors like:

```
/tmp/v_501/linked-list.6170371166411288392.tmp.c:12293:33: error: member reference type 'main__LinkedList *' (aka 'struct main__LinkedList *') is a pointer; did you mean to use '->'?
                _option_main__Node_ptr _t3 = l.tail;
```

and the message "This is a compiler bug, please report it" so I am doing so. I suspect the errors are caused by lines 21, 28 and 33, which all have the form:

`mut my_var := struct_var.reference_field or { panic("") }`

I'm not sure whether the fact that `reference_field` is *also* an `Option` is important.

File in question:

```
module main

struct Node {
	value int
mut:
	prev ?&Node = none
	next ?&Node = none
}

struct LinkedList {
mut:
	head ?&Node = none
	tail ?&Node = none
}

pub fn (mut l LinkedList) push(value int) {
	node := &Node{value:value, prev:l.tail}
	if l.head == none {
		l.head = node
	}	else {
		mut tail := l.tail or { panic("head is not none but tail is") }
		tail.next = node
	}
	l.tail = node
}

pub fn (mut l LinkedList) pop() int {
	tail := l.tail or { panic("Empty list not supported") }
	if tail.prev == none { // single element
		l.head = none
		l.tail = none
	}	else { // multiple elements
		mut prev := tail.prev or { panic("prev cannot be none at this point") }
		prev.next = none
		l.tail = prev
	}
	return tail.value
}
```


### Expected Behavior

Successful compilation.

### Current Behavior

```
% v -cg linked-list.v
/tmp/v_501/linked-list.6170371166411288392.tmp.c:2214:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/linked-list.6170371166411288392.tmp.c:12293:33: error: member reference type 'main__LinkedList *' (aka 'struct main__LinkedList *') is a pointer; did you mean to use '->'?
                _option_main__Node_ptr _t3 = l.tail;
                                             ~^
                                              ->
/tmp/v_501/linked-list.6170371166411288392.tmp.c:12311:32: error: member reference type 'main__LinkedList *' (aka 'struct main__LinkedList *') is a pointer; did you mean to use '->'?
        _option_main__Node_ptr _t1 = l.tail;
                                     ~^
                                      ->
/tmp/v_501/linked-list.6170371166411288392.tmp.c:12325:36: error: member reference type 'main__Node *' (aka 'struct main__Node *') is a pointer; did you mean to use '->'?
                _option_main__Node_ptr _t4 = tail.prev;
                                             ~~~~^
                                                 ->
1 warning and 3 errors generated.
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```


### Reproduction Steps

1. `v linked-list.v`

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 0bd0942

### Environment details (OS name and version, etc.)

OS: macos, macOS, 13.2.1, 22D68
Processor: 10 cpus, 64bit, little endian, Apple M1 Max
CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)

getwd: /Users/liteyear/Programming/Exercism/vlang/exercises/practice/linked-list
vmodules: /Users/liteyear/.vmodules
vroot: /Users/liteyear/Programming/v
vexe: /Users/liteyear/Programming/v/v
vexe mtime: 2023-03-12 22:12:47
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 9c9adb1.0bd0942

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: weekly.2023.10-41-g0bd09429
.git/config present: true
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -547,10 +547,11 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 				g.write('/* autoref */&')
 			}
 
-			if sfield.expected_type.has_flag(.option)
+			if (sfield.expected_type.has_flag(.option) && !sfield.typ.has_flag(.option))
 				|| (sfield.expected_type.has_flag(.result) && !sfield.typ.has_flag(.result)) {
 				g.expr_with_opt(sfield.expr, sfield.typ, sfield.expected_type)
 			} else {
+				g.left_is_opt = true
 				g.expr_with_cast(sfield.expr, sfield.typ, sfield.expected_type)
 			}
 		}

--- a/vlib/v/tests/option_struct_init_with_ref_opt_test.v
+++ b/vlib/v/tests/option_struct_init_with_ref_opt_test.v
@@ -1,0 +1,51 @@
+struct Node {
+	value int
+mut:
+	prev ?&Node = none
+	next ?&Node = none
+}
+
+struct LinkedList {
+mut:
+	head ?&Node = none
+	tail ?&Node = none
+}
+
+pub fn (mut l LinkedList) push(value int) {
+	node := &Node{
+		value: value
+		prev: l.tail
+	}
+	if l.head == none {
+		l.head = node
+	} else {
+		mut tail := l.tail or { panic('head is not none but tail is') }
+		tail.next = node
+	}
+	l.tail = node
+}
+
+pub fn (mut l LinkedList) pop() int {
+	tail := l.tail or { panic('Empty list not supported') }
+	if tail.prev == none { // single element
+		l.head = none
+		l.tail = none
+	} else { // multiple elements
+		mut prev := tail.prev or { panic('prev cannot be none at this point') }
+		prev.next = none
+		l.tail = prev
+	}
+	return tail.value
+}
+
+fn test_option_struct_init_with_ref_opt() {
+	mut ll := LinkedList{}
+	ll.push(11)
+	ll.push(22)
+	ret1 := ll.pop()
+	println(ret1)
+	assert ret1 == 22
+	ret2 := ll.pop()
+	println(ret2)
+	assert ret2 == 11
+}

